lib/environment/background/redirection.js

```diff
@@ -0,0 +1,114 @@
+/* @flow */
+
+import once from 'lodash/once';
+import keyBy from 'lodash/keyBy';
+
+import type { RedirectRule } from '../utils/redirection';
+import { requiredPermissions } from '../utils/redirection';
+import { _get as get, _set as set } from '../utils/storage';
+import { addListener } from './messaging';
+import { checkPermissionsGranted } from './permissions';
+
+let redirectRules: { [string]: RedirectRule };
+
+init();
+
+async function init() {
+	// pick up rules from storage so redirects work on cold start
+	redirectRules = deserializeRedirectRules(await get('RESRedirect', []));
```

Comment:
I made the rules persistent because leaving the responsibility of initializing the background redirection script's to foreground scripts (like the redirect module) causes redirect rules to not be applied on "cold start", such as when a user goes directly to a redirect-from page (ex: bookmarked reddit.com) after just starting the browser.

The other option (other than leaving responsibility to fg scripts) was to make the redirection bg script pick up option values from various other modules' storage, and then process them into redirection rules. So the bg script would need to know which other modules' options to pick up *and* how to translate their values to RedirectRules, which makes it less self-contained.

Redirects continuing to apply after a backup-restore that should disable them seems like a pretty serious problem. Perhaps we can solve it with something like this:
1. Redirection bg script keeps track of which fg modules supplied the redirection rules in its (persistent) state (can just extend the RedirectionRule type to have moduleID)
2. When restoring a backup causes a fg module's enabled-state to change, a message is sent to the background scripts
3. Redirection bg script listens for this message and removes all rules supplied by the fg module that was turned off

The opposite case (backup restore turning a disabled module back on), we can just have fg modules send their rules to the bg script onBeforeLoad

I'm not familiar with the code handling backup-restoration yet, so maybe the above doesn't make sense. Will look into how it works in a bit.